homotopy-core/src/graph.rs

```diff
@@ -1,310 +1,295 @@
+    Ix2: IndexType,
 {
-    use Height::{Regular, Singular};
-
-    let mut exploded_graph: SliceGraph<NewNodeKey, NewEdgeKey, T> = Default::default();
-
-    // Maps every node in the original graph to its slices in the exploded graph.
-    let mut node_to_slices: IdxVec<NodeIndex<S>, Vec<Option<NodeIndex<T>>>> =
-        IdxVec::with_capacity(graph.node_count());
-
-    for (n, (key, diagram)) in graph.node_references() {
-        let diagram: &DiagramN = diagram.try_into()?;
-
-        let mut slices = Vec::with_capacity(diagram.size() * 2 + 3);
-
-        // Source slice
-        slices.push((
-            node_map(n, Boundary::Source.into(), key.clone()),
-            diagram.source(),
-        ));
-
-        // Interior slices
-        for (i, slice) in diagram.slices().enumerate() {
-            slices.push((node_map(n, Height::from(i).into(), key.clone()), slice));
-        }
-
-        // Target slice
-        slices.push((
-            node_map(n, Boundary::Target.into(), key.clone()),
-            diagram.target(),
-        ));
-
-        let nodes: Vec<Option<NodeIndex<T>>> = slices
-            .into_iter()
-            .map(|(k, d)| k.map(|i| exploded_graph.add_node((i, d))))
-            .collect();
-
-        // Identity rewrite from source slice
-        if let (Some(s), Some(t)) = (nodes[0], nodes[1]) {
-            if let Some(key) = edge_map(None, RewriteOrigin::Boundary(Boundary::Source), None) {
-                exploded_graph.add_edge(s, t, (key, Rewrite::identity(diagram.dimension() - 1)));
-            }
-        }
+    pub output: SliceGraph<V, E, Ix2>,
+    pub node_to_nodes: IdxVec<NodeIndex<Ix1>, Vec<NodeIndex<Ix2>>>,
+    pub node_to_edges: IdxVec<NodeIndex<Ix1>, Vec<EdgeIndex<Ix2>>>,
+    pub edge_to_edges: IdxVec<EdgeIndex<Ix1>, Vec<EdgeIndex<Ix2>>>,
+}
 
-        // Rewrites between interior slices
-        for (i, cospan) in diagram.cospans().iter().enumerate() {
-            if let Some(singular) = nodes[usize::from(Singular(i)) + 1] {
-                if let Some(regular) = nodes[usize::from(Regular(i)) + 1] {
-                    if let Some(key) =
-                        edge_map(None, RewriteOrigin::Internal(i, Direction::Forward), None)
-                    {
-                        exploded_graph.add_edge(regular, singular, (key, cospan.forward.clone()));
-                    }
-                }
+impl<V, E, Ix1, Ix2> Deref for ExplosionOutput<V, E, Ix1, Ix2>
+where
+    Ix1: IndexType,
+    Ix2: IndexType,
+{
+    type Target = SliceGraph<V, E, Ix2>;
 
-                if let Some(regular) = nodes[usize::from(Regular(i + 1)) + 1] {
-                    if let Some(key) =
-                        edge_map(None, RewriteOrigin::Internal(i, Direction::Backward), None)
-                    {
-                        exploded_graph.add_edge(regular, singular, (key, cospan.backward.clone()));
-                    }
-                }
-            }
-        }
+    fn deref(&self) -> &Self::Target {
```

Comment:
You can use it to fake inheritance without trait objects. This use-case is not explicitly mentioned in the docs but is used _everywhere_ in Rust.

Reply:
Although I'm not sure I see the point of new-typing this and then immediately writing a `Deref` and `DerefMut` impl - maybe just a type alias?

Reply:
I want explode to return a few things and complex tuples are discouraged (clippy used to warn about this until I disabled it), so that's what I wrote the struct. The reason I implemented Deref is so you can chain explode calls without having to extract the output.

Reply:
`type SliceGraph = ...`

Reply:
Hmm, that's certainly an option. I will consider it.

Reply:
Actually I don't think I like that idea, `explode` has a very complicated type signature and I would rather not have two definitions of it that we have to keep in sync.

Reply:
I've made some changes - have a look and feel free to reverse the commit if you hate it